### PR TITLE
Add docs for Painter & Controller

### PR DESCRIPTION
**Pull request by cmyr**
This is the first part of the major section going over custom
widgets and the general widget model.

### CODE REVIEW

File: docs/book_examples/src/custom_widgets_md.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            self.suppress_next = !self.suppress_next;
+            if self.suppress_next {
+                return;
+            }
+        }
+
+        // if we want our child to receive this event, we must send it explicitly.
+        child.event(ctx, event, data, env);
+    }
+}
```
Comment (Finnerale) [inline docs/book_examples/src/custom_widgets_md.rs:59]: This is what came to my mind based on #787 .
A `TextBox` that fires an action (e.g. for search) when hitting enter or 300 ms after the last key press.
```rust
const ACTION: Selector = Selector::new("hello.textbox-action");

struct TextBoxActionController {
    timer: Option<TimerToken>,
}

impl TextBoxActionController {
    pub fn new() -> Self {
        TextBoxActionController { timer: None }
    }

    // Fire ACTION after 300 ms
    fn deadline() -> Instant {
        Instant::now() + Duration::from_millis(300)
    }
}

impl Controller<String, TextBox> for TextBoxActionController {
    fn event(
        &mut self,
        child: &mut TextBox,
        ctx: &mut EventCtx,
        event: &Event,
        data: &mut String,
        env: &Env,
    ) {
        match event {
            Event::KeyDown(k) if k.key_code == KeyCode::Return => {
                ctx.submit_command(ACTION, None);
            }
            Event::KeyUp(k) if k.key_code != KeyCode::Return => {
                self.timer = Some(ctx.request_timer(Self::deadline()));
                child.event(ctx, event, data, env);
            }
            Event::Timer(token) if Some(*token) == self.timer => {
                ctx.submit_command(ACTION, None);
            }
            _ => child.event(ctx, event, data, env),
        }
    }
}
```

Comment (cmyr) [reply docs/book_examples/src/custom_widgets_md.rs]: thanks, I've changed the example to this.

File: docs/book_examples/src/custom_widgets_md.rs [hunk 2]
```diff
@@ -32,12 +36,25 @@ fn background_label() -> impl Widget<Color> {
...
-impl Controller<String, TextBox> for AnnoyingController {
+impl TextBoxActionController {
+    pub fn new() -> Self {
+        TextBoxActionController { timer: None }
+    }
+
+    // Fire ACTION after 300 ms
+    fn deadline() -> Instant {
+        Instant::now() + DELAY
+    }
```
Comment (Finnerale) [inline docs/book_examples/src/custom_widgets_md.rs:54]:
```suggestion
```

File: docs/book_examples/src/custom_widgets_md.rs [hunk 3]
```diff
@@ -46,15 +63,19 @@ impl Controller<String, TextBox> for AnnoyingController {
...
-        if matches!(event, Event::KeyDown(k) if k.key_code == KeyCode::Backspace) {
-            self.suppress_next = !self.suppress_next;
-            if self.suppress_next {
-                return;
+        match event {
+            Event::KeyDown(k) if k.key_code == KeyCode::Return => {
+                ctx.submit_command(ACTION, None);
+            }
+            Event::KeyUp(k) if k.key_code != KeyCode::Return => {
+                self.timer = Some(ctx.request_timer(Self::deadline()));
```
Comment (Finnerale) [inline docs/book_examples/src/custom_widgets_md.rs:71]:
```suggestion
                self.timer = Some(ctx.request_timer(Instant::now() + DELAY));
```

**Pull request merged**